bitwarden_access_manager.py

```diff
@@ -0,0 +1,737 @@
+            f'member_name = "{self.name}"',
+            f'email = "{self.email}"',
+            f'type = "{self.type.name.lower()}"',
+            f"access_all = {str(self.access_all).lower()}",
+        ]
+        return "\n".join(lines)
+
+
+class GroupMember(NamedTuple):
+    member_id: int
+    member_name: str
+    group_name: str
+
+    def get_id(self) -> str:
+        # Our generic differ has the ability to turn add/removes into changes
+        # for pairs with the same id, but this does not apply to memberships,
+        # which do not themselves have an id, so the identity to group on is
+        # the value itself.
+        return f"{self.member_id}@{self.group_name}"
+
+    def format_toml(self) -> str:
+        # Needed to satisfy Diffable, but not used in this case.
+        raise Exception(
+            "Group memberships are not expressed in toml, "
+            "please print the diffs in some other way."
+        )
+
+
+class Group(NamedTuple):
+    id: str
+    name: str
+    access_all: bool
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> Group:
+        access_all: bool = False
+        if "access_all" in data:
+            access_all = data["access_all"]
+
+        return Group(
+            id=data["group_id"],
+            name=data["group_name"],
+            access_all=access_all,
+        )
+
+    def format_toml(self) -> str:
+        lines = [
+            "[[group]]",
+            f'group_id = "{self.id}"',
+            f'group_name = "{self.name}"',
+            f'access_all = "{str(self.access_all).lower()}"',
+        ]
+        return "\n".join(lines)
+
+
+class MemberCollectionAccess(NamedTuple):
+    name: str
+    access: GroupAccess
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> MemberCollectionAccess:
+        return MemberCollectionAccess(
+            name=data["member_name"],
+            access=GroupAccess[data["access"].upper()],
+        )
+
+    def format_toml(self) -> str:
+        return (
+            '{ member_name = "'
+            + self.name
+            + '", access = "'
+            + self.access.name.lower()
+            + '"}'
+        )
+
+
+class GroupCollectionAccess(NamedTuple):
+    name: str
+    access: GroupAccess
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> GroupCollectionAccess:
+        return GroupCollectionAccess(
+            name=data["group_name"],
+            access=GroupAccess[data["access"].upper()],
+        )
+
+    def format_toml(self) -> str:
+        return (
+            '{ group_name = "'
+            + self.name
+            + '", access = "'
+            + self.access.name.lower()
+            + '" }'
+        )
+
+
+class Collection(NamedTuple):
+    id: str
+    external_id: str
+    group_access: Optional[Tuple[GroupCollectionAccess, ...]]
+    member_access: Optional[Tuple[MemberCollectionAccess, ...]]
```

Comment:
```suggestion
    group_access: Tuple[GroupCollectionAccess, ...]
    member_access: Tuple[MemberCollectionAccess, ...]
```
In this case there is no semantic difference between `None` and an empty tuple, so we can simplify this a bit and drop the `Optional`.

Reply:
https://github.com/ChorusOne/github-access-manager/pull/10/commits/cda04f02c7fcdfb6a3af943476a052c413fb6493